ratis-common/src/main/java/org/apache/ratis/util/AwaitToRun.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Use a {@link Daemon},
+ * which repeatedly waits for a signal to run a method.
+ * <p>
+ * This class is threadsafe.
+ *
+ * @see AwaitForSignal
+ */
+public class AwaitToRun implements AutoCloseable {
+  public static final Logger LOG = LoggerFactory.getLogger(AwaitToRun.class);
+
+  private class RunnableImpl implements Runnable {
+    private final Runnable runMethod;
+
+    private RunnableImpl(Runnable runMethod) {
+      this.runMethod = runMethod;
+    }
+
+    @Override
+    public void run() {
+      for (; ; ) {
+        try {
+          awaitForSignal.await();
+        } catch (InterruptedException e) {
+          LOG.info("{} is interrupted", awaitForSignal);
+          Thread.currentThread().interrupt();
+          return;
+        }
+
+        try {
+          runMethod.run();
```

Comment:
It looks like the thread will never sleep after the first signal, even if there is no write, it will always be busy executing the run function. I'm not sure if I've got it wrong

Reply:
If I understand correctly, once the `runMethod` (e.g. `checkAndEvictCache`) finishes, it will loop back to `awaitForSignal.await` which will block until `signal()` is sent.

Also, I think even that this ensures `checkAndEvictCache` run sequentially. If `cacheEviction.signal()` is sent multiple times for each append entry, `checkAndEvictCache` only runs it if `runMethod` finished running (i.e. `awaitForSignal.await()`. So this might avoid the call for `checkAndEvictCache` for every append entry.

Please correct me if I'm wrong.

Reply:
Sorry, it's my fault! I was wondering why [getAndSet](https://github.com/apache/ratis/blob/master/ratis-common/src/main/java/org/apache/ratis/util/AwaitForSignal.java#L72) was used here, now I understand, very clever design!

Reply:
> So this might avoid the call for checkAndEvictCache for every append entry.

Yes, it definitely reduces the number of calls ~ so that leaves the concurrency control part above to see how we can do it ~